### PR TITLE
chore: clean up code in mermaidAPI render() and write specs/tests

### LINKED ISSUE #3648: (chore) DRY, clean up mermaidAPI, create tests
**Describe the bug**
_This is a chore, not a bug.  But I don't know what the best way is to present this._

1. Clean up the code in mermaidAPI: DRY up, refactor, use constants, etc.
Breaking up the giant functions will mean we can write tests.
2. Write specs/tests for mermaidAPI

This will make maintenance easier (code will be easier to comprehend, especially with the specs) and more reliable (because of the tests in the specs).

**This should _not_ change any functionality at all.**  

I'd like to do this before doing #3626 .

### PR DESCRIPTION
**Pull request by weedySeaDragon**
## :bookmark_tabs: Summary

**Cleaned up  and refactored `render()` in `mermaidAPI`** so that specs/tests could  be written for discrete steps, reduced duplication, and made it easier to understand for future maintainers.

Partially resolves #3648 

## :straight_ruler: Design Decisions

There is more refactoring and clean-up that can be done (including the other main functions), but this is a first pass. `render()` seemed the most important because it is the core functionality.

**Could not figure out how to mock/spyOn just one function within mermaidAPI, so some tests  instead painfully check results.  See TODOs in mermaidAPI.spec.ts.**

- cleaned up code: DRY, created constants for literals, renamed variables to clarify intent
- refactored into testable functions 
- wrote specs/tests 

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

**Comment by sidharthv96** [comment]
@weedySeaDragon, added some more types and slightly wandered into some related files. 
Saw some TODOs, are they intended for a discussion here?

**Comment by weedySeaDragon** [comment]
@sidharthv96 -- looks like I shouldn't change `if` statements to one line. Ex: 
```
if (someConditional) {
  something = someNewValue; // some short 1 line statement
}
```
changed to
```
if (someConditional)  something = someNewValue; // some short 1 line statement
```

There is an eslint rule for this: `curly: "error"`  (default is that all should be blocks with curly braces).  Should we put it into .eslintrc.json?  (Obv we should create a separate issue.  Mentioning it here because I was looking at your changes.

**Comment by weedySeaDragon** [comment]
@sidharthv96 Re TODOs -- mostly they are for the future.  But we can discuss and try to resolve with this PR if you think that's best.

**Comment by sidharthv96** [comment]
> There is an eslint rule for this: `curly: "error"` (default is that all should be blocks with curly braces). Should we put it into .eslintrc.json? (Obv we should create a separate issue. Mentioning it here because I was looking at your changes.

I didn't add curly to keep the footprint low. We should add it separately. I'll create an issue.

**Comment by knsv** [comment]
@weedySeaDragon @sidharthv96 Hey, whats the status on this one?

**Comment by weedySeaDragon** [comment]
I've been waiting for more reviews.

(I'll merge it with the most recent to bring it up to date.)

File: cSpell.json [hunk 1]
```diff
@@ -6,6 +6,7 @@
     "customizability",
     "Gantt",
     "jison",
+    "mermaid",
```
Comment (sidharthv96) [inline cSpell.json:9]: The lint error is for `mermiad's`. 
Isn't mermaid already an english word? Why do we need that in cspell.json?

Comment (weedySeaDragon) [reply cSpell.json:9]: After you clean up cSpell.json in #3753 , I'll update this. (adding "mermaid" into cSpell was my bad. I was kind of blindly correcting "mermiad")

Comment (weedySeaDragon) [reply cSpell.json:9]: @sidharthv96  Actually, looking at the failure, it's not eslint that is failing.  
It's failing in **`docs:verify`**

**Comment by weedySeaDragon** [comment]
@sidharthv96 - This is still failing during `pnpm run docs:verify`.
The `package.json` file doesn't even have any of the `docs..` scripts.  (They got removed during all of the vitepress/documentation changes.)  
Even so, when I (manually) run
```
ts-node-esm packages/mermaid/src/docs.mts
```
no document are updated ("_Transforming 0 ___ files.._." for all file types), and

```
ts-node-esm packages/mermaid/src/docs.mts --verify
```
verifies 0 files (for all file types).

Not sure how else to verify or change the `src/docs` files locally to get this to pass during CI.

**Comment by sidharthv96** [comment]
I think this might be the issue.

**Review by sidharthv96 — APPROVED** [review]
LGTM!

This should make the merge from master with the duplicate renderAsync much smaller.

**Review by MermaidChart — APPROVED** [review]
Good job! I can see how the d3 mock can be usefull!

**Pull request merged**